awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -0,0 +1,408 @@
+
+        response_body = ''
+        try:
+            response_body = response.read()
+        except(Exception) as e:
+            self.fail_json(msg="Failed to read response body: {0}".format(e))
+
+        response_json = {}
+        if response_body and response_body != '':
+            try:
+                response_json = loads(response_body)
+            except(Exception) as e:
+                self.fail_json(msg="Failed to parse the response json: {0}".format(e))
+
+        return {'status_code': response.status, 'json': response_json}
+
+    def authenticate(self, **kwargs):
+        if self.username and self.password:
+            # Attempt to get a token from /api/v2/tokens/ by giving it our username/password combo
+            # If we have a username and password we need to get a session cookie
+            login_data = {
+                "description": "Ansible Tower Module Token",
+                "application": None,
+                "scope": "write",
+            }
+            # Post to the tokens endpoint with baisc auth to try and get a token
+            api_token_url = (self.url._replace(path='/api/v2/tokens/')).geturl()
+
+            try:
+                response = self.session.open(
+                    'POST', api_token_url,
+                    validate_certs=self.verify_ssl, follow_redirects=True,
+                    force_basic_auth=True, url_username=self.username, url_password=self.password,
+                    data=dumps(login_data), headers={'Content-Type': 'application/json'}
+                )
+            except(Exception) as e:
+                # Sanity check: Did the server send back some kind of internal error?
+                self.fail_json(msg='Failed to get token: {0}'.format(e))
+
+            token_response = None
+            try:
+                token_response = response.read()
+                response_json = loads(token_response)
+                self.oauth_token_id = response_json['id']
+                self.oauth_token = response_json['token']
+            except(Exception) as e:
+                self.fail_json(msg="Failed to extract token information from login response: {0}".format(e), **{'response': token_response})
+
+        # If we have neiter of these then we can try un-authenticated access
+        self.authenticated = True
+
+    def default_check_mode(self):
+        '''Execute check mode logic for Ansible Tower modules'''
+        if self.check_mode:
+            try:
+                result = self.get_endpoint('ping')
+                self.exit_json(**{'changed': True, 'tower_version': '{0}'.format(result['json']['version'])})
+            except(Exception) as excinfo:
+                self.fail_json(changed=False, msg='Failed check mode: {0}'.format(excinfo))
+
+    def update_if_needed(self, existing_item, new_item, handle_response=True, **existing_return):
```

Comment:
Before this gets too far into the weeds, I want to ask about the highest priority design item, which is the consistency of how these methods are called.

It would be really good to distinguish all methods that terminate execution - that is they do exit_json or fail_json. I would also like to be very careful that they either always terminate or never terminate. You've done an excellent job fitting a solution to do that with the on_changed callback. Whether they terminate or not is so important, I almost think it should be in the method name.

On to the call arguments, the core CRUD actions involve `update_if_needed`, `post_endpoint`, and `delete_endpoint`. Consider the arguments for post_endpoint versus this method here.

```python
def post_endpoint(self, endpoint, handle_return=True, item_type='item', item_name='', *args, **kwargs):
```

This just doesn't look the same at all. My position is that these should all have a common set of functionality - because they are terminating methods.

Likewise, for deletion, you handle the case of changed=False in the module code very elegantly, but that's not consistent with this. If we have a method `update_if_needed`, then we should have a `delete_if_needed` method so that we know that (a) it will terminate execution and (b) it could terminate with multiple types of statuses.

I come back around to the design that tower-cli used, maybe we should have a single method for CRU and another one for D. Since the module itself deals with the `state` parameter, this makes a lot of sense with the surface area that Ansible requires.

Reply:
Great feed back.

In thinking about, I'm learning towards having C, U, and D methods. In the form of:
`update_if_needed`
`create_if_needed`
`delete_if_needed`

I don't want to put `_and_exit` on the functions since you can override the exit with `handle_response=False` and my gut is saying we will find some use case where we might have a "compound" action we need to take. I.e. create item 'a' and then create item 'b' which has a reference to 'a'. So the first one would handle_exit=False and the second one is True.

Each of the methods above can have a corresponding hook for `on_create`, `on_update`, and `on_delete`. Do you think those should be "registered" like I did or just make them parameters to the if_needed methods?

The R will continue to be handled via get_one and get_all (I'm thinking we can rename `get_all_endpoint` to just `get_all` for consistency).

In addition, I think it would make sense to merged CU method called `create_or_update_if_needed` so that modules can get into a very simple pattern like:
```
if state == 'absent':
    delete_if_needed(<args>)
else:
    create_or_update_if_needed(<args>)
```
Another other option would be to make a single method like: `put_object_in_state(old_object, new_object, state, ...)` which would handle all of the CUD operations in one. Any thoughts on that? That would put even the conditional logic into the module_util so the modules would only need to:
  * resolve associated IDs
  * build the object data structures
  * define any on_* methods 
  * make the put_object_in_state call.

Reply:
I do feel partial to the CU, and D methods, exactly how your proposed code snippet is. But as long as we have consistent args with calls to `_if_needed` methods, then I give my 👍 

> my gut is saying we will find some use case where we might have a "compound" action we need to take. I.e. create item 'a' and then create item 'b' which has a reference to 'a'. So the first one would handle_exit=False and the second one is True.

Good point, we should be prepared to complex multi-action scenarios. Given that, `handle_exit=False` sounds like a good solution to me... probably better than the callback approach. If the actions got complex, then we would wind up in callback hell. The way `tower_project` waits for update could be re-written to use a `handle_exit=False` kwarg and not the callback.

Reply:
Check it out now. I left the option for a callback for "simple" things, like waiting for a project update to complete. For more complex things you can give it handle_response=False but then you are on the hook for processing the return and follow up actions.

Reply:
I'm on board with the call pattern in the modules now. Having both the callback and handle_response kwarg has a pretty good justification.